lib/wallaroo/core/sink/connector_sink/connector_sink_checkpoint_rollback.pony

```diff
@@ -0,0 +1,465 @@
+
+    @ll(debug, "CpRbTransition:: %s -> %s".cstring(),
+      curr.name().cstring(), next.name().cstring())
+    // We must update sink's _cprb pointer before calling .enter()!
+    // Otherwise, if .enter() also updates the _cprb pointer, then
+    // a pointer update will get clobbered & lost.
+    sink._update_cprb_member(next)
+    next.enter(sink)
+    @ll(debug, "CpRbTransition:: %s -> %s DONE".cstring(),
+      curr.name().cstring(), next.name().cstring())
+
+class _CpRbAbortCheckpoint is _CpRbOps
+  let _debug: LogSevCat = Log.make_sev_cat(Log.debug(), Log.conn_sink())
+  var _checkpoint_to_abort: (None | CheckpointBarrierToken)
+  var _abort_sent: Bool = false
+  let _send_phase1: Bool
+
+  fun name(): String => __loc.type_name()
+
+  new create(checkpoint_to_abort: (None | CheckpointBarrierToken)) =>
+    _checkpoint_to_abort = checkpoint_to_abort
+    _send_phase1 = match checkpoint_to_abort
+      | None =>
+        true
+      else
+        false
+      end
+
+  fun ref enter(sink: ConnectorSink ref) =>
+    sink.cprb_send_advertise_status(false)
+    match sink.cprb_check_sink_phase_partial_barrier()
+    | None =>
+      None
+    | let cbt: CheckpointBarrierToken =>
+      @l(Log.info(), Log.conn_sink(),
+        "_CpRbAbortCheckpoint: detected shear with token %s".cstring(),
+          cbt.string().cstring())
+      match _checkpoint_to_abort
+      | let cpta: CheckpointBarrierToken =>
+        if cpta != cbt then
+          @l(Log.crit(), Log.conn_sink(),
+            "Expected %s but partial barrier is %s".cstring(),
+            _checkpoint_to_abort.string().cstring(), cbt.string().cstring())
+          Fail()
+        end
+      end
+    end
+    _ChangeSinkPhaseQueueMsgsForwardTokens(sink where shear_risk = true)
+    _is_checkpoint_id_known(sink)
+
+  fun ref _is_checkpoint_id_known(sink: ConnectorSink ref) =>
+    match _checkpoint_to_abort
+    | let cbt: CheckpointBarrierToken =>
+      if not _abort_sent then
+        // Do not send 2PC commands here.  Though we can assume that
+        // ExtConn's TCP connection is still established, we don't
+        // actually know the connection's state. Downstream processing
+        // later will take care of all scenarios.
+        let txn_id = sink.cprb_make_txn_id_string(cbt.id)
+        sink.cprb_send_abort_to_barrier_coordinator(cbt, txn_id)
+        _abort_sent = true
+      end
+    end
+
+  fun ref checkpoint_complete(sink: ConnectorSink ref,
+    checkpoint_id: CheckpointId)
+  =>
+    // This checkpoint is complete, so there's no way that we can
+    // abort it later.  We must wait for the next checkpoint to
+    // start.
+    _checkpoint_to_abort = None
+
+  fun ref phase1_abort(sink: ConnectorSink ref, txn_id: String) =>
+    // This is in response to the 2PC Phase1 message that we (probably)
+    // sent in _is_checkpoint_id_known().
+    // Call _is_checkpoint_id_known(), just in case.
+    _is_checkpoint_id_known(sink)
```

Comment:
This seems like it should return a Bool. Can you rename to reflect what's happening?